# Notebook 1 — Read & Join Tables

## Setup — imports and database connection

In [3]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("postgresql://olist_user:olist_pass@localhost:5432/olist_db")

## Read all tables from the database

In [4]:
with engine.connect() as conn:
    orders = pd.read_sql("SELECT * FROM orders", conn)
    customers = pd.read_sql("SELECT * FROM customers", conn)
    order_items = pd.read_sql("SELECT * FROM order_items", conn)
    order_payments = pd.read_sql("SELECT * FROM order_payments", conn)
    products = pd.read_sql("SELECT * FROM products", conn)
    sellers = pd.read_sql("SELECT * FROM sellers", conn)

## Check the shape of each table

In [5]:
print("orders:", orders.shape)
print("customers:", customers.shape)
print("order_items:", order_items.shape)
print("order_payments:", order_payments.shape)

orders: (99441, 8)
customers: (99441, 5)
order_items: (112650, 7)
order_payments: (103886, 5)


**Note:** `order_items` (112,650 rows) and `order_payments` (103,886 rows) 
have more rows than `orders` (99,441 rows) — meaning some orders have 
more than one item or one payment. These tables need aggregation before joining.

## Check for duplicate rows per order (why we need to aggregate)

In [6]:
print("orders rows:", len(orders), "| unique order_id:", orders["order_id"].nunique())
print("order_items rows:", len(order_items), "| unique order_id:", order_items["order_id"].nunique())
print("order_payments rows:", len(order_payments), "| unique order_id:", order_payments["order_id"].nunique())

orders rows: 99441 | unique order_id: 99441
order_items rows: 112650 | unique order_id: 98666
order_payments rows: 103886 | unique order_id: 99440


**Note:** `orders` has one row per order_id, confirmed (99,441 = 99,441). 
But `order_items` has 98,666 unique orders out of 112,650 rows, and 
`order_payments` has 99,440 unique orders out of 103,886 rows — confirming 
duplication that needs to be handled before merging.

## Aggregate order_items to one row per order

In [7]:
items_agg = order_items.groupby("order_id").agg(
    n_items=("order_item_id", "count"),
    total_price=("price", "sum"),
    total_freight=("freight_value", "sum"),
    avg_price=("price", "mean"),
    n_unique_sellers=("seller_id", "nunique"),
    n_unique_products=("product_id", "nunique"),
).reset_index()

print(items_agg.shape)
items_agg.head()

(98666, 7)


,order_id,n_items,total_price,total_freight,avg_price,n_unique_sellers,n_unique_products
0,00010242fe8c5a6d1ba2dd792cb16214,1,58.90,13.29,58.90,1,1
1,00018f77f2f0320c557190d7a144bdd3,1,239.90,19.93,239.90,1,1
2,000229ec398224ef6ca0657da4fc703e,1,199.00,17.87,199.00,1,1
3,00024acbcdf0a6daa1e931b038114c75,1,12.99,12.79,12.99,1,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,199.90,18.14,199.90,1,1


**Note:** After aggregation, `items_agg` has exactly one row per order, 
with useful summary columns like total price, number of items, and 
number of unique sellers.

## Aggregate order_payments to one row per order

In [8]:
payments_agg = order_payments.groupby("order_id").agg(
    total_payment_value=("payment_value", "sum"),
    n_payment_installments=("payment_installments", "max"),
    n_payment_methods=("payment_type", "nunique"),
).reset_index()

payments_agg.head()

,order_id,total_payment_value,n_payment_installments,n_payment_methods
0,00010242fe8c5a6d1ba2dd792cb16214,72.19,2,1
1,00018f77f2f0320c557190d7a144bdd3,259.83,3,1
2,000229ec398224ef6ca0657da4fc703e,216.87,5,1
3,00024acbcdf0a6daa1e931b038114c75,25.78,2,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,218.04,3,1


**Note:** Same idea — `payments_agg` now has one row per order with 
total payment value and payment method info.

## Merge everything into one ML table (one row per order)

In [9]:
ml_table = orders.merge(customers, on="customer_id", how="left")
ml_table = ml_table.merge(items_agg, on="order_id", how="left")
ml_table = ml_table.merge(payments_agg, on="order_id", how="left")

print("Final shape:", ml_table.shape)
print("Unique order_id:", ml_table["order_id"].nunique())

Final shape: (99441, 21)
Unique order_id: 99441


**Note:** The final `ml_table` has the same number of rows as `orders` 
(99,441), and unique order_id count matches too — confirming the merge 
did not create duplicate rows.

## Save the artifact

In [10]:
ml_table.to_csv("../artifacts/01_ml_table.csv", index=False)
print("Saved:", ml_table.shape)

Saved: (99441, 21)
